xtask/src/main.rs

```diff
@@ -0,0 +1,589 @@
+use std::{
+    collections::{HashMap, HashSet},
+    fs,
+    io::Read,
+    path::Path,
+    process::{Command, Stdio},
+};
+
+use anyhow::{anyhow, Context, Result};
+use console::Style;
+use similar::{ChangeTag, TextDiff};
+use structopt::StructOpt;
+
+#[derive(Debug, StructOpt)]
+struct Options {
+    #[structopt(
+        long,
+        short,
+        help = "keep target toolchains that were installed as dependency"
+    )]
+    keep_targets: bool,
+
+    #[structopt(
+        long,
+        short,
+        help = "treat compiler warnings as errors (RUSTFLAGS=\"--deny warnings\")"
+    )]
+    deny_warnings: bool,
+
+    #[structopt(subcommand)]
+    cmd: TestCommand,
+}
+
+#[derive(StructOpt, Debug)]
+#[allow(clippy::enum_variant_names)]
+enum TestCommand {
+    TestAll,
+    TestBook,
+    TestCross,
+    TestHost,
+    TestLint,
+    TestSnapshot,
+}
+
+fn run_command<P: AsRef<Path>>(
+    cmd_and_args: &[&str],
+    cwd: Option<P>,
+    env: &[(&str, &str)],
+) -> Result<()> {
```

Comment:
As far as I can see the generic parameter is actually not needed. (You need to adapt all calls to `run_command` and remove `::<&str>`).

```suggestion
fn run_command(cmd_and_args: &[&str], cwd: Option<&str>, env: &[(&str, &str)]) -> Result<()> {
```

Reply:
yeah, I mean I wanted to have the flexibility to pass `cwd`s in any useful form, but when I made `env` generic too that became too cumbersome, so I took *that* out again ... in the end we're probably fine with no generics here.

Reply:
and we can add that if we actually need it. and if we don't add it we probably don't need it 🤷🏾‍♂️